File: CPP/Queue.cpp
#include <iostream>
using namespace std;

struct node
{
    int value;
    node *next;
};

node *createNode(int value)
{
    node *n1 = new node;
    n1->value = value;
    n1->next = NULL;
    return n1;
}

class Queue
{
    node *front, *rear;

public:
    Queue() { front = rear = NULL; }
    void enQueue(int value)
    {
        node *newNode = createNode(value);
        if (front == NULL)
        {
            front = rear = newNode;
            return;
        }
        rear->next = newNode;
        rear = newNode;
    }
    int deQueue()
    {
        if (front == NULL)
        {
            cout << "Underflow\n";
            return NULL;
        }
        int temp = front->value;
        if (front == rear)
            front = rear = NULL;
        else
            front = front->next;
        return temp;
    }
    void displayAll()
    {
        if (front == NULL)
        {
            cout << "Underflow\n";
            return;
        }
        node *ptr = front;
        while (ptr)
        {
            cout << ptr->value << "->";
            ptr = ptr->next;
        }
        cout << endl;
    }
};

// Test Code

int main()
{
    Queue Q;
    Q.enQueue(1);
    Q.enQueue(2);
    Q.enQueue(3);
    Q.displayAll();
    Q.deQueue();
    Q.displayAll();
    Q.deQueue();
    Q.displayAll();
    return 0;
}

File: Python/Data_Structures/Queue.py
class Node(object):
    def __init__(self, value = None):
        self.value = value
        self.next = None

class Queue(object):
    def __init__(self):
        self.front = self.rear = None

    def enQueue(self, value):
        newNode = Node(value)
        if self.front == None:
            self.front = self.rear = newNode
            return
        self.rear.next = newNode
        self.rear = newNode

    def deQueue(self):
        if self.front == None:
            raise ValueError('Underflow\n')
        nodeDel = self.front
        if self.front == self.rear:
            self.front = self.rear = None
            return nodeDel.value
        self.front = self.front.next
        return self.front.value

    def displayAll(self):
        if self.front == None:
            raise ValueError('Underflow\n')
        ptr = self.front
        while ptr != None:
            print(ptr.value, end=' ')
            ptr = ptr.next
        print()
    
# Test Code
Q = Queue()
Q.enQueue(1)
Q.enQueue(2)
Q.enQueue(3)
Q.enQueue(4)
Q.displayAll()
Q.deQueue()
Q.deQueue()
Q.displayAll()